File: csharp/1687.delivering-boxes-from-storage-to-ports.cs
/*
 * @lc app=leetcode id=1687 lang=csharp
 *
 * [1687] Delivering Boxes from Storage to Ports
 */

// @lc code=start
public class Solution {
    public int BoxDelivering(int[][] boxes, int portsCount, int maxBoxes, int maxWeight) {
        int n = boxes.Length, ports = 0, weight = 0, i = 0;
        var dp = new int[n + 1];
        for (int j = 0; j < n; ++j) {
            weight += boxes[j][1];
            if (j > 0 && boxes[j][0] != boxes[j - 1][0]) {
                ++ports;
            }
            while (j - i >= maxBoxes || weight > maxWeight || i < j && dp[i] == dp[i + 1]) {
                weight -= boxes[i][1];
                if (boxes[i][0] != boxes[i + 1][0]) {
                    --ports;
                }
                ++i;
            }
            dp[j + 1] = dp[i] + ports + 2;
        }
        return dp[n];
    }
}
// @lc code=end

File: java/1687.delivering-boxes-from-storage-to-ports.java
/*
 * @lc app=leetcode id=1687 lang=java
 *
 * [1687] Delivering Boxes from Storage to Ports
 */

// @lc code=start
class Solution {
    public int boxDelivering(int[][] boxes, int portsCount, int maxBoxes, int maxWeight) {
        int n = boxes.length, w = 0, l = 0, p = 0;
        int[] dp = new int[n + 1];
        for (int r = 0; r < n; ++r) {
            w += boxes[r][1];
            if (r > 0 && boxes[r][0] != boxes[r - 1][0]) {
                ++p;
            }
            while (w > maxWeight || r - l >= maxBoxes || (l < r && dp[l] == dp[l + 1])) {
                w -= boxes[l][1];
                if (boxes[l][0] != boxes[l + 1][0]) {
                    --p;
                }
                ++l;
            }
            dp[r + 1] = p + 2 + dp[l];
        }
        return dp[n];
    }
}
// @lc code=end

